### PR TITLE
Add calls to set annotation for kubectl apply

### PR DESCRIPTION
**Pull request by jackgr**
This pull request adds calls to create configuration annotations used by kubectl apply when configuration is updated through most code paths. The only configuration changes it does not capture are those made outside kubectl, and those made by kubectl patch, since the user does not supply the new configuration in its entirety for a patch.

This should be the final step in implementing #1702. Note that it depends on and by necessity includes #14621, which adds the method to create the three way merge patch used by kubectl apply. When #14621 is merged, the diff for this pull request will be significantly smaller.

cc @bgrant0607, @ghodss, @JanetKuo, @mikedanese

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -64,6 +64,10 @@ func AddSourceToErr(verb string, source string, err error) error {
 
 var fatalErrHandler = fatal
 
+// ApplyAnnotation is the annotation used to store the previous configuration
+// of a resource that can be used in a three way diff by UpdateApplyAnnotation.
+const ApplyAnnotation = "k8s-kubectl-apply"
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/helpers.go]: Should be "kubectl.kubernetes.io/apply" or similar

File: pkg/kubectl/cmd/util/helpers.go [hunk 2]
```diff
@@ -400,22 +404,70 @@ func DumpReaderToFile(reader io.Reader, filename string) error {
...
+	// If there weren't any, then make a new map of annotations.
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// Retrieve the original configuration from the annotations, if present.
+	original := []byte(annotations[ApplyAnnotation])
+
+	var modified []byte
+	if info.VersionedObject != nil {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/helpers.go]: Should this reset the apply annotation on the versioned object before serializing it?  It seems like this would add the annotation every edit recursively, eventually leading to exceeding the max annotation size.

Comment (jackgr) [inline pkg/kubectl/cmd/util/helpers.go]: Good question, and one that I thought about and addressed. The solution here is subtle, and probably deserves a comment. The VersionedObject is the input to the command, not the object from the server. It therefore doesn't contain the annotation unless the user has put it there. I thought about stripping it out if they did, but that seems contrary to the philosophy of letting the user define the desired behavior. Most users will never think to add the annotation on their own. Those who do will do so consciously, and will observe that it gets nested. However, unless they add it back to their input the next time around, it will never nest more than one deep.

So, bottom line... no, it doesn't recurse or even embed the previous annotation by default, but the user can achieve one or multiple levels of embedded annotations if they explicitly set it up that way.

File: pkg/kubectl/cmd/util/helpers.go (hunk 1 above)
Comment (jackgr) [inline pkg/kubectl/cmd/util/helpers.go]: I looked for precedent, and found only the labels used by add-ons, which I emulated. Do we have examples of annotations or labels which use the pattern you're recommending? If not, why is it better than the pattern used by add-on labels?

Comment (jackgr) [inline pkg/kubectl/cmd/util/helpers.go]: Agreed per [this comment](https://github.com/kubernetes/kubernetes/pull/14621#discussion_r40510257), which identifies precedents.

Comment (jackgr) [inline pkg/kubectl/cmd/util/helpers.go]: Done in #14621 on which this PR depends and by necessity includes.

**Comment by jackgr** [comment]
Rebased to #14621.

File: pkg/util/strategicpatch/patch.go [hunk 3]
```diff
@@ -827,3 +838,104 @@ func sliceElementType(slices ...[]interface{}) (reflect.Type, error) {
...
+	return err == ErrConflict
+}
+
+// IsPreconditionFailed returns true if the provided error indicates
+// a Delta precondition did not succeed.
+func IsPreconditionFailed(err error) bool {
+	return err == ErrPreconditionFailed
+}
+
+var ErrPreconditionFailed = fmt.Errorf("a precondition failed")
```
Comment (kargakis) [inline pkg/util/strategicpatch/patch.go]: Can you group all the error types in this file together? Also do we want them exported or not? Half of them are not, these are...

Comment (jackgr) [inline pkg/util/strategicpatch/patch.go]: The ones that are exported are the only ones that should be exported, and they are grouped together.

**Comment by jackgr** [comment]
@bgrant0607, @ghodss Added e2e test. PTAL.

**Comment by jackgr** [comment]
@jlowdermilk Changes integrated and rebased. PTAL.

**Comment by jlowdermilk** [comment]
lgtm
